### PR TITLE
Converts to NPM scripts, removes Gulp

### PR DESCRIPTION
**Pull request by bigtimebuddy**
## Overview

This change refactors the build process to remove Gulp and Gulp-related plugins. This simplifies the build process mostly through the **package.json** and a few helper scripts. It has the same feature parity as the Gulp tasking (e.g., custom builds by ignore modules), but with less devDependencies overhead and without requiring the global `gulp` install. 

### Changes

* `gulp build` is now `npm run build`
* `gulp watch` is now `npm start`
* `gulp jshint` is now `npm run lint`
* `gulp clean` is now `npm run clean`
* Excluding changed from `gulp build -e filters` to `npm run build -- -e filters`
* `gulp docs` is removed, `npm run docs` still works

### Fixed

* License header is re-added to both the debug and release builds.

### CODE REVIEW

**Comment by bigtimebuddy** [comment]
@englercj I welcome your feedback here.

File: .travis.yml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 sudo: false
 language: node_js
 node_js:
-    - "0.10"
+    - "0.12"
```
Comment (englercj) [inline .travis.yml]: Might as well go big and do v4 or v6. The pre-v1 versions aren't supported by the node team anymore.

File: package.json [hunk 2]
```diff
@@ -16,10 +16,16 @@
...
-    "test": "gulp && testem ci",
-    "build": "gulp",
-    "docs": "jsdoc -c ./gulp/util/jsdoc.conf.json -R README.md"
+    "clean": "rimraf bin && mkdirp bin",
+    "prestart": "npm run clean",
+    "start": "parallelshell 'npm run watch:debug' 'npm run watch:release'",
+    "watch:debug": "watchify src/index.js -s PIXI -o bin/pixi.js -dv",
+    "watch:release": "watchify src/index.js -s PIXI -o bin/pixi.min.js -dv",
+    "test": "npm run build && testem ci",
+    "lint": "jshint --reporter=scripts/reporter.js scripts/*.js src/*.js src/**/*.js",
```
Comment (englercj) [inline package.json]: Shouldn't need `src/*.js`, should be covered by the `src/**/*.js` glob.

File: scripts/bundle.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        }
+        catch(e){
+            console.log('> ERROR: Module not found for \'%s\'', exclude);
+        }
+    });
+    console.log('');
+}
+
+// Build the debug version
+execSync('browserify ' + exclusions + 'src -s PIXI -d | exorcist bin/pixi.js.map > bin/pixi.js');
```
Comment (englercj) [inline scripts/bundle.js]: Might as well just use the browserify api isntead of the CLI since your already in a script.

File: scripts/bundle.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+// Fix the output 
+postbundle('bin/pixi.js', true);
```
Comment (englercj) [inline scripts/bundle.js]: modifying the bundle file after the fact will make the sourcemaps incorrect. If you append the header after the bundle is done, the sourcemaps will have incorrect line numbers.

File: scripts/bundle.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+// Build the release version
+execSync('browserify ' + exclusions + 'src -s PIXI -d | exorcist bin/pixi.min.js.map | uglifyjs -cm > bin/pixi.min.js');
```
Comment (englercj) [inline scripts/bundle.js]: Recommend just using the node APIs of these modules so you can build once and then write it out, then minify that in memory without having to bundle twice (that will take quite some time).

File: scripts/postbundle.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    license = license.replace('${date}', (new Date()).toString())
+        .replace('${version}', packageInfo.version);
+
+    // Add the sourcemapurl back to minified
+    if (!debug) {
+        contents += '//# sourceMappingURL=' + path.basename(bundle) + '.map';
+    }
+
+    // Re-write the file
+    fs.writeFileSync(bundle, license + contents);
```
Comment (englercj) [inline scripts/postbundle.js]: See above comments about adding lines to the start of the bundle.

**Comment by englercj** [comment]
:+1: Good luck!

**Comment by ivanpopelyshev** [comment]
Oh my god! Shall I remove gulp from pixi-spine and other plugins?

**Comment by bigtimebuddy** [comment]
Okay @englercj have another crack at it. Completely uses the Browserify API this time without a hacky post process.

**Comment by bigtimebuddy** [comment]
I couldn't find something that _exactly_ did the license header insert in the way I wanted with Browserify, so I created it as a plugin.

**Comment by bigtimebuddy** [comment]
@ivanpopelyshev I don't think you need to, at least not right way. I did, however, create a Gulp-less build process for [pixi-filters](https://github.com/pixijs/pixi-filters).

**Comment by ivanpopelyshev** [comment]
This is counter-revolution! Npm strikes back!

**Comment by englercj** [comment]
Neat, looks good to me

**Comment by bigtimebuddy** [comment]
Hey @GoodBoyDigital, care to weight in?

**Comment by bigtimebuddy** [comment]
Also, would like to sanity check if someone could try to build this branch on Windows.

**Comment by ivanpopelyshev** [comment]
"scripts" is not a command

```txt
D:\pixi-games\pixi.js>npm run build

> pixi.js@4.0.0 prebuild D:\pixi-games\pixi.js
> npm run lint && npm run clean


> pixi.js@4.0.0 lint D:\pixi-games\pixi.js
> jshint --reporter=scripts/reporter.js scripts src


√ No problems


> pixi.js@4.0.0 clean D:\pixi-games\pixi.js
> rimraf bin && mkdirp bin


> pixi.js@4.0.0 build D:\pixi-games\pixi.js
> scripts/bundle.js

"scripts" не является внутренней или внешней командой, исполняемой программой или пакетным файлом.
```

File: package.json [hunk 7]
```diff
@@ -16,10 +16,16 @@
...
-    "docs": "jsdoc -c ./gulp/util/jsdoc.conf.json -R README.md"
+    "clean": "rimraf bin && mkdirp bin",
+    "prestart": "npm run clean",
+    "start": "parallelshell 'npm run watch:debug' 'npm run watch:release'",
+    "watch:debug": "watchify src/index.js -s PIXI -o bin/pixi.js -dv",
+    "watch:release": "watchify src/index.js -s PIXI -o bin/pixi.min.js -dv",
+    "test": "npm run build && testem ci",
+    "lint": "jshint --reporter=scripts/reporter.js scripts src",
+    "prebuild": "npm run lint && npm run clean",
+    "build": "scripts/bundle.js",
```
Comment (englercj) [inline package.json]: `node scripts/bundle.js`

**Comment by bigtimebuddy** [comment]
@ivanpopelyshev what if you change the `"build"` script in **package.json** to be `"node scripts/bundle.js"` instead of just `"scripts/bundle.js"`

**Comment by ivanpopelyshev** [comment]
It works.

Now we have to fix "npm start"

```txt
> pixi.js@4.0.0 start D:\pixi-games\pixi.js
> parallelshell 'npm run watch:debug' 'npm run watch:release'

"'npm" не является внутренней или внешней
командой, исполняемой программой или пакетным файлом.
```

The ' symbol is breaking it somehow

**Comment by bigtimebuddy** [comment]
@ivanpopelyshev What about escaping double quotes: `"parallelshell \"npm run watch:debug\" \"npm run watch:release\""`

**Comment by ivanpopelyshev** [comment]
Wow! IT works! Awesome!

**Comment by ivanpopelyshev** [comment]
It removes gulp client, right?

**Comment by ivanpopelyshev** [comment]
Is it ok ?

```js
var uglify = require('gulp-uglify');
```

at the same time, you added uglify-js into dependencies

**Comment by bigtimebuddy** [comment]
@ivanpopelyshev Yup, I removed uglify-js. Good catch.

gulp-uglify, gulp-sourcemaps and gulp-empty should be in the **package.json**. These are plugins for piping streams and not just for Gulp. For instance they work with browserify's piping.

**Comment by GoodBoyDigital** [comment]
looking good @bigtimebuddy 👍

**Comment by GoodBoyDigital** [comment]
nice work dude 👍

**Comment by mreinstein** [comment]
this is fantastic. One less dependency/build system i need to grok. :)

**Pull request merged**